astrolabe/spec_runner.py

```diff
@@ -0,0 +1,400 @@
+
+    def is_cluster_state(self, goal_state):
+        cluster_info = self.cluster_url.get().data
+        return cluster_info.stateName.lower() == goal_state.lower()
+
+    def verify_cluster_configuration_matches(self, state):
+        """Verify that the cluster config is what we expect it to be (based on
+        maintenance status). Raises AssertionError."""
+        state = state.lower()
+        if state not in ("initial", "final"):
+            raise AstrolabeTestCaseError(
+                "State must be either 'initial' or 'final'.")
+        cluster_config = self.cluster_url.get().data
+        assert_subset(
+            cluster_config,
+            self.spec.maintenancePlan[state].clusterConfiguration)
+        process_args = self.cluster_url.processArgs.get().data
+        assert_subset(
+            process_args, self.spec.maintenancePlan[state].processArgs)
+
+    def initialize(self):
+        """
+        Initialize a cluster with the configuration required by the test
+        specification.
+        """
+        LOGGER.info("Initializing cluster {!r}".format(self.cluster_name))
+
+        cluster_config = self.spec.maintenancePlan.initial.\
+            clusterConfiguration.copy()
+        cluster_config["name"] = self.cluster_name
+        try:
+            self.client.groups[self.get_group().id].clusters.post(
+                **cluster_config)
+        except AtlasApiError as exc:
+            if exc.error_code == 'DUPLICATE_CLUSTER_NAME':
+                # Cluster already exists. Simply re-configure it.
+                # Cannot send cluster name when updating existing cluster.
+                cluster_config.pop("name")
+                self.client.groups[self.get_group().id].\
+                    clusters[self.cluster_name].patch(**cluster_config)
+
+        # Apply processArgs if provided.
+        process_args = self.spec.maintenancePlan.initial.processArgs
+        if process_args:
+            self.client.groups[self.get_group().id].\
+                clusters[self.cluster_name].processArgs.patch(**process_args)
+
+    def run(self, persist_cluster=False):
+        LOGGER.info("Running test {!r} on cluster {!r}".format(
+            self.id, self.cluster_name))
+
+        # Step-0: sanity-check the cluster configuration.
+        self.verify_cluster_configuration_matches("initial")
+
+        # Start the test timer.
+        timer = Timer()
+        timer.start()
+
+        # Step-1: load test data.
+        test_data = self.spec.driverWorkload.get('testData')
+        if test_data:
+            LOGGER.info("Loading test data on cluster {!r}".format(
+                self.cluster_name))
+            connection_string = self.get_connection_string()
+            client = MongoClient(connection_string, w="majority")
+            coll = client.get_database(
+                self.spec.driverWorkload.database).get_collection(
+                self.spec.driverWorkload.collection)
+            coll.drop()
```

Comment:
do you need to potentially catch `ns not found`? or is that not a problem for python?

Reply:
PyMongo ignores this error on `drop()` so no special handling is required.